playwright/tests/common/environment.spec.ts

```diff
@@ -26,15 +51,42 @@ test.describe("Environment", () => {
   test.run()(
     'I can see "Production" environment when no environments exist',
     async ({ sliceMachinePage }) => {
+      await mockManagerProcedures({
+        page: sliceMachinePage.page,
+        procedures: buildMockEnvironmentProcedures({
+          environments: environments.slice(0, 1),
+        }),
+      });
+
+      await sliceMachinePage.gotoDefaultPage();
+      await expect(
+        sliceMachinePage.menu.environmentSelector.environmentName,
+      ).toHaveText("Production");
+    },
+  );
+
+  test.run()(
+    'I can see "Production" environment when an invalid environment is set',
+    async ({ sliceMachinePage }) => {
+      let didMockFetchActiveEnvironmentOnce = false;
       await mockManagerProcedures({
         page: sliceMachinePage.page,
         procedures: [
+          ...buildMockEnvironmentProcedures({ environments }),
           {
-            path: "prismicRepository.fetchEnvironments",
-            data: () => ({
-              environments: environments.slice(0, 1),
-            }),
-            execute: false,
+            path: "prismicRepository.fetchActiveEnvironment",
+            data: (data) => {
+              if (didMockFetchActiveEnvironmentOnce) {
```

Comment:
My other proposal was to handle an array of procedures where you can define twice the same procedure path. Like that if you define it 2 times, the first one will be the first call, the second one will be the second call + all next. WDYT?
See: https://github.com/prismicio/slice-machine/blob/53b09cc308e1fe1d8e816b1025833cabaf1a2a78/playwright/utils/mockManagerProcedures.ts

Reply:
Something like MSW's fall-through behavior would be nice. It's similar to what you are describing, but a little more flexible and natural-feeling.

Basically, if a route interceptor returns a `Response` instance, the response is used. If a `Response` isn't returned, the next handler for that route is used. It continues until a `Response` is found, or fails to handle the request if none are there. The handlers can be set to only be used once.

We can't do that exact system in our case, but we can do something similar:

- Maintain a list of handlers.
- Every time `mockManagerProcedures()` is called, the handlers are added to the start of the list.
- The first matching handler is used.
- Handlers can be configured to be removed after being used once.

This allows us to set up procedure handlers at the start of the test and add/modify new ones throughout the test. It also allows us to mock behavior that happens without user interaction (like the above `didMockFetchActiveEnvironmentOnce` logic).

This improvement would also remove the need for the `buildMockEnvironmentProcedures()` helper I wrote, which was needed to cut down on the repetitiveness and verbosity.

Reply:
It means we can write tests like this:

```ts
  test.run()(
    "I can see the window top border depending on the environment",
    async ({ sliceMachinePage }) => {
      await mockManagerProcedures({
        page: sliceMachinePage.page,
        procedures: [
          {
            path: "prismicRepository.fetchEnvironments",
            data: () => ({ environments }),
            execute: false,
          },
        ],
      });

      await sliceMachinePage.gotoDefaultPage();

      await mockManagerProcedures({
        page: sliceMachinePage.page,
        procedures: [
          {
            path: "prismicRepository.fetchActiveEnvironment",
            data: () => ({ activeEnvironment: environments[0] }),
            execute: false,
          },
        ],
      });
      await sliceMachinePage.menu.environmentSelector.selectEnvironment(
        environments[0].name,
      );
      await expect(sliceMachinePage.topBorder).toHaveCSS(
        "background-color",
        "rgb(109, 84, 207)",
      );

      await mockManagerProcedures({
        page: sliceMachinePage.page,
        procedures: [
          {
            path: "prismicRepository.fetchActiveEnvironment",
            data: () => ({ activeEnvironment: environments[1] }),
            execute: false,
          },
        ],
      });
      await sliceMachinePage.menu.environmentSelector.selectEnvironment(
        environments[1].name,
      );
      await expect(sliceMachinePage.topBorder).toHaveCSS(
        "background-color",
        "rgb(56, 91, 204)",
      );

      await mockManagerProcedures({
        page: sliceMachinePage.page,
        procedures: [
          {
            path: "prismicRepository.fetchActiveEnvironment",
            data: () => ({ activeEnvironment: environments[2] }),
            execute: false,
          },
        ],
      });
      await sliceMachinePage.menu.environmentSelector.selectEnvironment(
        environments[2].name,
      );
      await expect(sliceMachinePage.topBorder).toHaveCSS(
        "background-color",
        "rgb(255, 159, 26)",
      );
    },
  );
```

Reply:
@xrutayisire and I discussed it off GitHub and come to the conclusion that an API like this would be ideal:

```ts
procedures.mock("project.fetchActiveEnvironment", ({ data, args }) => {
  // ...
}, {
  execute: false,
  times: 1
})
```

`procedures` would be a fixture with a stateful list of procedures for the test. As `mock()` is called, handlers are added to the list.

It will be implemented in a separate PR and ported to these tests.